pycbc/psd/variation.py

```diff
@@ -210,10 +210,197 @@ def find_trigger_value(psd_var, idx, start, sample_rate):
+
+    The Live strain data is convolved with the filter to produce a timeseries
+    containing the PSD variation values for each sample. This mean square of
+    the timeseries is then taken over the short_stride to remove the effects of
+    short duration glitches and further outliers from the mean are replaced
+    within the array. This array is then further averaged every second to
+    produce a timeseries that will contain a number of value equal to the
+    increment.
+
+    Parameters
+    ----------
+    strain : pycbc.timeseries
+        Live data being searched through by the PyCBC Live search.
+    full_filt : some sort of array, scipy so i guess numpy?
+        A filter created by `live_create_filter`.
+    increment : float
+        The number of seconds in each increment in the PyCBC Live search.
+    data_trim : float
+        The number of seconds to be trimmed from either end of the convolved
+        timeseries to prevent artefacts.
+    short_stride : float
+        The number of seconds to average the PSD variation timeseries over to
+        remove the effects of short duration glitches.
+
+    Returns
+    -------
+    psd_var : pycbc.timeseries
+        A timeseries containing the PSD variation values for each second
+        an increments number of seconds prior to the end of the input strain
+        data.
+
+    """
+    sample_rate = int(strain.sample_rate)
+
+    # Grab the last increments worth of data with extra to account for the need
+    #  to trim the data to remove edge effects.
+    astrain = strain.time_slice(strain.end_time - increment - (data_trim*3),
+                                strain.end_time)
+
+    # Convole the data and the filter to produce the PSD variation timeseries,
+    #  then trim the beginning and end of the data to prevent edge effects.
+    wstrain = sig.fftconvolve(astrain, full_filt, mode='same')
+    wstrain = wstrain[int(data_trim * sample_rate):-int(data_trim * sample_rate)]
+
+    # Create a PSD variation array by taking the mean square of the PSD
+    #  variation timeseries every short_stride
+    short_ms = numpy.mean(
+        wstrain.reshape(-1, int(sample_rate * short_stride)) ** 2, axis=1)
+
+    # Create an array of averages to substitute out outliers in the PSD
+    #  variation array
+    ave = 0.5 * (short_ms[2:] + short_ms[:-2])
+    outliers = short_ms[1:-1] > (2. * ave)
+    short_ms[1:-1][outliers] = ave[outliers]
+
+    # Calculate the average of the PSD variation array for every second
+    m_s = []
+    stride = 1 / short_stride
+    for idx in range(int(len(short_ms) / stride)):
+        start = int(stride * idx)
+        end = int(stride * (idx + 1))
+        m_s.append(numpy.mean(short_ms[start:end]))
+
+
+    # Convert m_s to a numpy array
+    m_s = numpy.array(m_s, dtype=wstrain.dtype)
+
+    # Convert the m_s numpy array to a pycbc timeseries which now contains the
+    #  psd variation value every second.
+    psd_var = TimeSeries(m_s,
```

Comment:
However, I am not sure why the 'epoch' given here is correct. Why does this look significantly different from the offline code?

Reply:
From looking at the offline code, it is generating PSD var values for the whole 512 seconds of data and so it is logical to start from strain.start_time when defining the new timeseries.

In the live code we only want to find PSD var values for the newest increment of strain data so taking the end time of the data and subtracting the increment and the extra data from trimming is how I found the epoch of the psd var timeseries.

Example:
start_time = 0
end_time = 512
increment = 8
data_trim = 2

We take the latest data from 504 - 512, pad the start by 6 seconds and trim 2s from both sides. We calculate the mean square and average each second so we have 10 data points to put into our timeseries and we need to line the epoch up which is: end_time - increment - 2\*data_trim.

Reply:
OK, so the calculation is done starting with data from 498 through 512, 2s are lost for filter artefacts giving 500 through 510, then the epoch is 1 'trim segment' later than the start of the original data, i.e. 2 rather than 3 'trims' plus the analysis stride ('increment').

Reply:
The epoch is 1 trim segment later than the start of the slice of the data we are using to calculate PSD variation, not the start of the original data because that varies depending on how much data PyCBC Live has in the buffer right now.

I could change the epoch to be `epoch=astrain.start_time + data_trim` if you would prefer? The change mainly being from strain -> astrain where astrain is the sliced strain data.

Reply:
I resolved the conversation as I didn't think there was a problem in the end.  But maybe it would be good to add a short comment to explain the time accounting, eg `The original data segment started at 'increment + 3 * data_trim' before the strain end_time, the variation time series starts 1 * data_trim later`